### ISSUE TITLE
Issue when setting scatter color in separate method call

### ISSUE THREAD

**Issue by astrofrog** [matplotlib/matplotlib#10381]
For an application I am developing, I need to call ``ax.scatter`` then change the colors of the points after the fact. However, if the input x/y values contain NaN values, then setting the colors after the fact doesn't work properly. The following example:

```python
import matplotlib
matplotlib.use('Agg')

import numpy as np
import matplotlib.pyplot as plt

fig = plt.figure()
ax = fig.add_subplot(1, 1, 1, aspect='auto')

x = np.array([1, 2, 3])
y = np.array([2, np.nan, 4])
c = np.array([3, 4, 5])

s = ax.scatter(x, y, c=c, vmin=3, vmax=5, s=320, zorder=1, edgecolor='none')

s = ax.scatter(x, y, s=160, zorder=2, edgecolor='none')
s.set_array(c)
s.set_clim(3, 5)

s = ax.scatter([], [], s=40, zorder=3, edgecolor='none')
s.set_offsets(np.vstack([x, y]).transpose())
s.set_array(c)
s.set_clim(3, 5)

fig.savefig('test.png')

```

produces the following output:

![test](https://user-images.githubusercontent.com/314716/35814346-bdb682f0-0a8d-11e8-846e-07b7a9120827.png)

The top right point should be all yellow. However it isn't, because in the ``scatter`` call, invalid values are removed, so when I try and set the colors with ``set_array``, the color values are offset from the x/y values. In the final case, things work since the 'bad' values are masked from neither the offsets nor the colors.

Is there a way to achieve something along the lines of the middle example (setting the colors after the fact)

**Comment by tacaswell** [comment]
This is because we create the initial artist and strip out any points that have `nan` in them.  The new colors will then be offset from where you think they should be.

attn @efiring as you have been working around this code recently.

**Comment by efiring** [comment]
I think that adding the requested functionality will require writing a subclass of `PathCollection` that can keep track of the original inputs, and can repeat the mask- and nan-filtering operation every time a relevant property is updated.

**Comment by anntzer** [comment]
Do we actually need to filter out the nans (except for backcompatibility considerations)?  Aren't the renderers supposed to be able to handle nan data to start with anyways?

**Comment by efiring** [comment]
Good question; maybe they all can.  If so, taking advantage of that would require some substantial changes, I think, because we support masked arrays, and use them internally.  Going 100% with nans for missing values implies always using floating point, and switching from `.max` to `.nanmax` etc.

**Comment by efiring** [comment]
I think this is closed by #12422.  If not, please reopen.
